Add test cases to verify blitFramebuffer with multiple drawbuffers

Richard-Yunchao opened the pull request:
@kenrussell , PTAL. Thanks a lot!

Add conformance test for this TODO: https://cs.chromium.org/chromium/src/gpu/command_buffer/service/gles2_cmd_decoder.cc?q=gles2_cmd_decoder&sq=package:chromium&l=7734.

kenrussell commented:
Thanks @Richard-Yunchao for this test. Are you sure it is correct? I see it fail both on a MacBook Pro, which might be expected because it's running OpenGL 4.1 and may need more workarounds -- but also on Linux with an NVIDIA GPU running OpenGL 4.5. It also fails on Linux with ANGLE's OpenGL backend.

Richard-Yunchao commented:
Ah... I made a mistake. I should change the srgb to linear as expected linear color if the src image is srgb and the dst image is linear. 
Now the code is OK. It can pass on Linux desktop with OGL 4.5. But it fails on MacOSX with OGL 4.1.

PTAL. @kenrussell . 

(I updated the code at local, then revise the code and re-submit, I found that the commit will contain many commits those have been merged... So I used "git pull --force" to remove those commits...)

zhenyao commented on sdk/tests/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
```diff
@@ -0,0 +1,228 @@
+    for (var ii = 0; ii < filters.length; ++ii) {
+        for (var jj = 0; jj < drawbuffersFormats.length; ++jj) {
+            blitframebuffer_srgb_and_linear_drawbuffers(gl.SRGB8_ALPHA8, drawbuffersFormats[jj], filters[ii]);
+            blitframebuffer_srgb_and_linear_drawbuffers(gl.RGBA, drawbuffersFormats[jj], filters[ii]);
```
nit: maybe use sized format gl.RGBA8 to be consistent?

Richard-Yunchao replied:
done

kenrussell commented on sdk/tests/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
```diff
@@ -0,0 +1,228 @@
+
+    if (drawbuffersFormatMask & linearMask) {
+        gl.bindTexture(gl.TEXTURE_2D, tex_linear0);
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, size, size, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
```
Use sized internal format here too?

Richard-Yunchao replied:
done

Richard-Yunchao replied:
done

kenrussell commented on sdk/tests/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
```diff
@@ -0,0 +1,228 @@
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, size, size, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
+        gl.framebufferTexture2D(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT1, gl.TEXTURE_2D, tex_linear0, 0);
+        gl.bindTexture(gl.TEXTURE_2D, tex_linear1);
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, size, size, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
```
Here too?

Richard-Yunchao replied:
done

kenrussell commented on sdk/tests/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
```diff
@@ -0,0 +1,228 @@
+}
+
+function checkPixel(color, expectedColor) {
+  var tolerance = 7;
```
Does the tolerance need to be this high? This test passes on my MacBook Pro Retina with NVIDIA GPU with a tolerance of 2.

zhenyao replied:
This seems very large.  Maybe 3?

Richard-Yunchao replied:
I set it to 3 in the new commit.

Richard-Yunchao replied:
Done

Richard-Yunchao commented on sdk/tests/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
```diff
@@ -0,0 +1,228 @@
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "setup draw framebuffer should succeed");
+
+    var status = gl.checkFramebufferStatus(gl.DRAW_FRAMEBUFFER) && gl.checkFramebufferStatus(gl.READ_FRAMEBUFFER);
+    if (status != gl.FRAMEBUFFER_COMPLETE) {
```
Ah... My mistake... It is weird to me too. The js works...

zhenyao commented:
What about kbr's question where the updated test fails?

Richard-Yunchao commented:
@zhenyao and @kenrussell , the blitframebuffer doesn't honor the drawBuffers() API at my side (MacOSX with Intel GPU, OGL 4.1). Only the COLOR_ATTACHMENT0 is rendered during blitFramebuffer. 

I will ask driver team whether they have latest driver to update.

@kenrussell , I do a slightly revision, where the blitframebuffer must filter the image: src region and dst region is different now. The previous commits have the same bounds for src region and dst region, the driver may can do some optimization without filtering. So, could you retest the lastest commit on Mac Nvidia, @kenrussell ?

kenrussell commented:
LGTM

All of the tests pass on my Retina MacBook Pro with NVIDIA GPU and 10.11.6.

Unfortunately, you'll need to resolve conflicts before this can be merged. Thanks for the new test.

zhenyao commented:
LGTM

Richard-Yunchao commented:
Thanks for your review, @kenrussell and @zhenyao, I fixed the conflicts now. But It seems that the pr is in a mess... I updated my local master to upstream, then rebase master to this branch locally, fix the conflict and then push this branch.
